services/invitations/pkg/backends/keycloak/backend.go

```diff
@@ -0,0 +1,149 @@
+// Package keycloak offers an invitation backend for the invitation service.
+// TODO: Maybe move this outside of the invitation service and make it more generic?
+
+package keycloak
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strings"
+
+	"github.com/Nerzal/gocloak/v13"
+	"github.com/google/uuid"
+	"github.com/owncloud/ocis/v2/ocis-pkg/log"
+	"github.com/owncloud/ocis/v2/services/invitations/pkg/invitations"
+)
+
+const (
+	idAttr       = "OWNCLOUD_ID"
+	userTypeAttr = "OWNCLOUD_USER_TYPE"
+	userTypeVal  = "Guest"
+)
+
+var userRequiredActions = []string{"UPDATE_PASSWORD", "VERIFY_EMAIL"}
+
+// Backend represents the keycloak backend.
+type Backend struct {
+	logger       log.Logger
+	client       *gocloak.GoCloak
+	clientID     string
+	clientSecret string
+	clientRealm  string
+	userRealm    string
+}
+
+// New returns a new keycloak.Backend, with all the config options set.
+func New(
+	logger log.Logger,
+	baseURL, clientID, clientSecret, clientRealm, userRealm string,
+	insecureSkipVerify bool,
+) *Backend {
+	client := gocloak.NewClient(baseURL)
+	restyClient := client.RestyClient()
+	restyClient.SetTLSClientConfig(&tls.Config{InsecureSkipVerify: insecureSkipVerify})
+	return &Backend{
+		logger: log.Logger{
+			logger.With().
+				Str("invitationBackend", "keycloak").
+				Str("baseURL", baseURL).
+				Str("clientID", clientID).
+				Str("clientRealm", clientRealm).
+				Str("userRealm", userRealm).
+				Logger(),
+		},
+		client:       client,
+		clientID:     clientID,
+		clientSecret: clientSecret,
+		clientRealm:  clientRealm,
+		userRealm:    userRealm,
+	}
+}
+
+// CreateUser creates a user in the keycloak backend.
+func (b Backend) CreateUser(ctx context.Context, invitation *invitations.Invitation) (string, error) {
+	token, err := b.getToken(ctx)
+	if err != nil {
+		return "", err
+	}
+	u := uuid.New()
```

Comment:
creating a user with a UUID works? @micbar said he ran into problems when trying to post users via the graph api ... 

🤔 

ahhh but keycloak will directly brovision them in the LDAP backend. Got it!

Reply:
Yes, and keycloak needs some wrangling to allow this, it's a bit tedious and clumsy, but it works.